### PR TITLE
fix empty polygons in shapes layer

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR fixes some errors that were happening in the shapes layer when polygons with no inside, or paths all at the same point were added. Double-clicking to create a new polygon could lead to one of these errors, and is thought to be the problem identified in #223.

There is still one error that occurs in the `vispy` triangulate method when all the points are colinear

```python
from vispy.geometry import PolygonData
import numpy as np

data = np.array([[0, 0], [0, 1], [0, 2]])
vertices, triangles = PolygonData(vertices=data).triangulate()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-26-1b6df5c6de06> in <module>()
      3 
      4 data = np.array([[0, 0], [0, 1], [0, 2]])
----> 5 vertices, triangles = PolygonData(vertices=data).triangulate()

~/anaconda3/lib/python3.6/site-packages/vispy/geometry/polygon.py in triangulate(self)
    128 
    129         tri = Triangulation(self._vertices, edges)
--> 130         tri.triangulate()
    131         return tri.pts, tri.tris
    132 

~/anaconda3/lib/python3.6/site-packages/vispy/geometry/triangulation.py in triangulate(self)
    139             # get index along front that intersects pts[i]
    140             idx = 0
--> 141             while pts[front[idx+1], 0] <= pi[0]:
    142                 idx += 1
    143             pl = pts[front[idx]]

IndexError: list index out of range
```

I have not investigated this further, but we could either fix within vispy or add some test for it before passing the points to this function  

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] `examples/add_shapes.py`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
I have now added a simple colinearity test. This should fix all the known bugs around polygon drawing

File: napari/layers/_shapes_layer/shape_util.py [hunk 1]
```diff
@@ -255,6 +255,29 @@ def orientation(p, q, r):
     return val
 
 
+def is_colinear(points):
+    """Determines is a list of 2D points are colinear
```
Comment (kne42) [inline napari/layers/_shapes_layer/shape_util.py]:
```suggestion
    """Determines if a list of 2D points are colinear.
```

File: napari/layers/_shapes_layer/shape_util.py [hunk 2]
```diff
@@ -255,6 +255,29 @@ def orientation(p, q, r):
...
+
+    Parameters
+    -------
+    points : np.ndarray
```
Comment (kne42) [inline napari/layers/_shapes_layer/shape_util.py]:
```suggestion
    points : (N, 2) array
```

File: napari/layers/_shapes_layer/shape_util.py [hunk 3]
```diff
@@ -255,6 +255,29 @@ def orientation(p, q, r):
...
+        Nx2 array of points to be tested for colinearity
```
Comment (kne42) [inline napari/layers/_shapes_layer/shape_util.py]:
```suggestion
        Points to be tested for colinearity
```

File: napari/layers/_shapes_layer/shape_util.py [hunk 4]
```diff
@@ -255,6 +255,29 @@ def orientation(p, q, r):
...
+
+    Returns
+    -------
+    val : bool
+        True is all points are colinear, False otherwise.
+    """
+    if len(points) < 3:
+        return True
+
+    for p in points[2:]:
```
Comment (kne42) [inline napari/layers/_shapes_layer/shape_util.py:276]: what's the `[2:]` for?

**Review by kne42 — CHANGES_REQUESTED** [review]
left some comments on documentation, may also want to look into using `np.fromiter` since you create arrays from generators very frequently

File: napari/layers/_shapes_layer/shape_util.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/_shapes_layer/shape_util.py]: fixed

File: napari/layers/_shapes_layer/shape_util.py (hunk 3 above)
Comment (sofroniewn) [reply napari/layers/_shapes_layer/shape_util.py]: fixed

File: napari/layers/_shapes_layer/shape_util.py (hunk 2 above)
Comment (sofroniewn) [reply napari/layers/_shapes_layer/shape_util.py]: fixed - but will need to change in other places, worth doing in a follow-up refactor PR

File: napari/layers/_shapes_layer/shape_util.py (hunk 4 above)
Comment (sofroniewn) [reply napari/layers/_shapes_layer/shape_util.py:276]: The collinearity test takes three points, the first two are the first two in the list, and then the third is iterated through in the loop. I will add a comment

**Review by kne42 — APPROVED** [review]
LGTM!

**Pull request merged**